src/Analyzers/Core/Analyzers/RemoveRedundantEquality/AbstractRemoveRedundantEqualityDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeAnalysis.RemoveRedundantEquality
+{
+    internal abstract class AbstractRemoveRedundantEqualityDiagnosticAnalyzer
+        : AbstractBuiltInCodeStyleDiagnosticAnalyzer
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+
+        protected AbstractRemoveRedundantEqualityDiagnosticAnalyzer(ISyntaxFacts syntaxFacts)
+            : base(IDEDiagnosticIds.RemoveRedundantEqualityDiagnosticId,
+                   option: null,
+                   new LocalizableResourceString(nameof(AnalyzersResources.Remove_redundant_equality), AnalyzersResources.ResourceManager, typeof(AnalyzersResources)))
+        {
+            _syntaxFacts = syntaxFacts;
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterOperationAction(AnalyzeBinaryOperator, OperationKind.BinaryOperator);
+
+        private void AnalyzeBinaryOperator(OperationAnalysisContext context)
+        {
+            var operation = (IBinaryOperation)context.Operation;
+            if (operation.OperatorMethod is not null)
+            {
+                // We shouldn't report diagnostic on overloaded operator as the behavior can change.
+                return;
+            }
```

Comment:
Do I still need this condition given that I also have:

```
            if (rightOperand.Type.SpecialType is not SpecialType.System_Boolean ||
                leftOperand.Type.SpecialType is not SpecialType.System_Boolean)
            {
                return;
            }
```

I can't think of a case where both sides are bool and the operator is overloaded. I mean that operator overloading is applied to user-defined types, but can't be applied to the existing bool type AFAIK.

Reply:
i would keep this check.